File: src/lib/server/modules/mangareader/mangareader.ts
import type { SearchResult, Manga, Chapter } from '../../classes/interfaces';
import Module, { ModuleFlags } from '../../classes/Module';
import axios from 'axios';
import { load as cheerioload } from 'cheerio';
import { ImageUnscrambler } from './ImageInterceptor';

class MangaReader extends Module {
	id = 'mr';
	name = 'MangaReader';
	flags: ModuleFlags[] = [ModuleFlags.DISABLE_GLOBAL_SEARCH];

	search(query: string) {
		return new Promise(async (resolve: (value: SearchResult[]) => void) => {
			if (query.length < 1) {
				resolve([]);
				return;
			}

			const $ = cheerioload(
				(await axios.get(`https://mangareader.to/search?keyword=${query}`)).data
			);

			const results: SearchResult[] = [];
			$('#main-content > section > div.manga_list-sbs > div.mls-wrap')
				.find('div.item.item-spc')
				.each((_, el) => {
					const ele = $(el);
					const a = ele.find('div.manga-detail > h3 > a');
					const langs = ele.find('a.manga-poster > span').text();

					langs.split('/').forEach((l) => {
						results.push({ id: a.attr('href') + '|' + l, title: `[${l}] ${a.attr('title')}` });
					});
				});

			resolve(results);
		});
	}

	manga(mangaid: string) {
		return new Promise(async (resolve: (value: Manga) => void) => {
			const { mangaId, lang } = (() => {
				const p = mangaid.split('|');
				return {
					mangaId: p[0],
					lang: p[1].toLowerCase()
				};
			})();

			const $ = cheerioload((await axios.get(`https://mangareader.to${mangaId}`)).data);

			const chapters: Chapter[] = [];
			$('#' + lang + '-chapters')
				.find('a')
				.each((_, ch) => {
					const chap = $(ch);
					chapters.push({
						title: chap.attr('title') + '',
						id: chap.attr('href') + '',
						date: new Date(0)
					});
				});

			resolve({
				synopsis: $('#modaldesc > div > div > div.modal-body > div').text().trim(),
				author:
					$(
						'#ani_detail > div > div > div.anis-content > div.anisc-detail > div.sort-desc > div.anisc-info-wrap > div.anisc-info > div:nth-child(3)'
					)
						.text()
						.split('Authors:')[1] || 'unknown',
				artist: '',
				img:
					$('#ani_detail > div > div > div.anis-content > div.anisc-poster > div > img').attr(
						'src'
					) + '',
				chapters,
				sourceurl: `https://mangareader.to${mangaId}`
			});
		});
	}

	chapter(manga: string, id: string) {
		return new Promise(async (resolve: (value: string[]) => void) => {
			let $ = cheerioload((await axios.get(`https://mangareader.to${id}`)).data);
			const chapid = $('#wrapper').attr('data-reading-id') + '';

			$ = cheerioload(
				(
					await axios.get(
						`https://mangareader.to/ajax/image/list/chap/${chapid}?mode=horizontal&quality=high&hozPageSize=1`
					)
				).data.html + ''
			);

			const res: string[] = [];
			const images = $('#divslide > div.divslide-wrapper').find('.ds-image');
			images.each((i, el) => {
				if (i == images.length - 1) return;
				const ele = $(el);
				if (ele.hasClass('shuffled')) {
					res.push(this.imageProxyUrl('', '', ele.attr('data-url') + ''));
				} else {
					res.push(ele.attr('data-url') + '');
				}
			});

			resolve(res);

			// console.log(res)
		});
	}

	image(manga: string, chapter: string, id: string) {
		return new Promise(async (resolve: (value: Buffer) => void) => {
			// console.log(manga, chapter, id)

			const response = await axios.get(id, {
				responseType: 'arraybuffer'
			});

			const init = new Uint8Array(Buffer.from(response.data, 'binary'));

			const res = await new ImageUnscrambler().interceptResponse(init);

			resolve(res);
		});
	}
}

export default new MangaReader();

File: pkg/modules/mangareader/mangareader.go
package mangareader

import (
	"fmt"
	"image/jpeg"
	"io"
	"net/http"
	"net/url"
	"strings"

	"github.com/Minettyx/FoolslideProxy/pkg/types"
	"github.com/Minettyx/FoolslideProxy/pkg/utils"

	mangareader_unshuffle "github.com/Minettyx/mangareader.to-image-unshuffle"
	"github.com/PuerkitoBio/goquery"
)

type mangaReader struct {
	baseUrl string
}

var Module = mangaReader{
	baseUrl: "https://mangareader.to",
}

var _ types.Module = mangaReader{}

func (c mangaReader) Id() string {
	return "mr"
}
func (c mangaReader) Name() string {
	return "MangaReader"
}
func (c mangaReader) Flags() types.ModuleFlags {
	return types.ModuleFlags{types.DISABLE_GLOBAL_SEARCH}
}

func (c mangaReader) Search(query string) ([]types.SearchResult, error) {
	if len(query) < 1 {
		return []types.SearchResult{}, nil
	}

	res, err := http.Get(c.baseUrl + "/search?keyword=" + url.QueryEscape(query))
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	results := []types.SearchResult{}

	doc.Find("div.item-spc").Each(func(i int, s *goquery.Selection) {
		a := s.Find(".manga-name > a")
		title := a.Text()

		mid, _ := a.Attr("href")
		img, _ := s.Find("img").Attr("src")

		langs := strings.Split(s.Find(".manga-poster > span").Text(), "/")

		for _, v := range langs {
			results = append(results, types.SearchResult{
				Id:    mid + "|" + v,
				Title: "[" + v + "] " + title,
				Image: img,
			})
		}

	})

	return results, nil
}

func (c mangaReader) Manga(id string) (*types.Manga, error) {
	p := strings.Split(id, "|")
	if len(p) < 2 {
		return nil, fmt.Errorf("Language not found in manga id")
	}

	mangaid := p[0]
	lang := strings.ToLower(p[1])

	res, err := http.Get(c.baseUrl + mangaid)
	if err != nil {
		return nil, err
	}

	defer res.Body.Close()

	doc, err := goquery.NewDocumentFromReader(res.Body)
	if err != nil {
		return nil, err
	}

	author := ""

	doc.Find(".anisc-info > .item").Each(func(i int, s *goquery.Selection) {
		switch s.Find(".item-head").Text() {
		case "Authors:":
			s.Find("a").Each(func(i int, s *goquery.Selection) {
				author += strings.Replace(s.Text(), ",", "", 1) + ", "
			})
		}
	})
	if len(author) > 2 {
		author = author[:len(author)-2]
	}

	img, _ := doc.Find("img.manga-poster-img").Attr("src")

	chapters := []types.Chapter{}

	doc.Find("#" + lang + "-chapters > li").Each(func(i int, s *goquery.Selection) {
		chtit := s.Find(".name").Text()
		chid, _ := s.Find("a").Attr("href")

		chapters = append(chapters, types.Chapter{
			Title: chtit,
			Id:    chid,
		})
	})

	return &types.Manga{
		Synopsis:  doc.Find(".description").Text(),
		Author:    author,
		Artist:    "",
		Img:       img,
		Sourceurl: c.baseUrl + mangaid,
		Chapters:  chapters,
	}, nil
}

func (c mangaReader) Chapter(manga, id string) ([]string, error) {
	doc, err := utils.GetAndGoquery(c.baseUrl + id)
	if err != nil {
		return nil, err
	}

	chapid, _ := doc.Find("#wrapper").Attr("data-reading-id")

	type chResp struct {
		Html string `json:"html"`
	}

	var resp chResp
	err, _ = utils.GetAndJsonParse(c.baseUrl+"/ajax/image/list/chap/"+chapid+"?mode=horizontal&quality=high&hozPageSize=1", &resp)
	if err != nil {
		return nil, err
	}

	doc, err = goquery.NewDocumentFromReader(strings.NewReader(resp.Html))

	res := []string{}

	doc.Find(".ds-image").Each(func(i int, s *goquery.Selection) {
		imgurl, ok := s.Attr("data-url")

		if !ok {
			return
		}

		if s.HasClass("shuffled") {
			res = append(res, "local://"+imgurl)
		} else {
			res = append(res, imgurl)
		}
	})

	return res, nil
}

func (c mangaReader) Image(manga, chapter, id string, w io.Writer) error {
	println(manga, chapter, id)

	origin, err := http.Get(id)
	if err != nil {
		return err
	}

	defer origin.Body.Close()

	input_image, err := jpeg.Decode(origin.Body)
	if err != nil {
		panic(err)
	}

	output_image := mangareader_unshuffle.Unshuffle(input_image)

	err = jpeg.Encode(w, output_image, nil)
	if err != nil {
		return err
	}
	return nil
}

func (c mangaReader) Latest() ([]types.LatestResult, error) {
	return []types.LatestResult{}, nil
}

func (c mangaReader) Popular() ([]types.PopularResult, error) {
	return []types.PopularResult{}, nil
}
